src/nvrtc/safe.rs

```diff
@@ -222,6 +225,257 @@ impl CompileOptions {
+    /// Returns an error if there is a problem cleaning the `PtxCrate`.
+    pub fn clean(&mut self) -> Result<(), String> {
+        match &self.project {
+            PtxCrateKind::Cargo { project_dir } => {
+                let manifest_path = project_dir.join("Cargo.toml");
+                let output = Command::new("cargo")
+                    .arg("clean")
+                    .arg("--manifest-path")
+                    .arg(manifest_path)
+                    .output()
+                    .map_err(|e| format!("Failed to execute command: {}", e))?;
+                if output.status.success() {
+                    self.kernels = None;
+                    Ok(())
+                } else {
+                    Err(format!(
+                        "Failed to build PTX file: {}",
+                        String::from_utf8_lossy(&output.stderr)
+                    ))
+                }
+            },
+            PtxCrateKind::_Standalone { standalone: _standalone } => 
+                Err("standalone RS to PTX not supported".to_string()), //todo!("rm -rf target/ ?"),
+        }
+    }
+    
+    /// Builds the PTX code for the currently set kernels.
+    ///
+    /// # Arguments
+    ///
+    /// * `self` - A mutable reference to the `PtxCrate` instance.
+    ///
+    /// # Returns
+    ///
+    /// Returns a reference to a vector of `Ptx` instances containing the compiled PTX code
+    /// for the kernels if the build is successful, otherwise an error message as a `String`.
+    ///
+    /// # Errors
+    ///
+    /// This method will return an error if the build process fails.
+    pub fn build_ptx(&mut self) -> Result<&Vec<Ptx>, String> {
+        match &self.project {
+            PtxCrateKind::Cargo { project_dir } => {
+                let manifest_path = project_dir.join("Cargo.toml");
+
+                let output = Command::new("cargo")
+                    .arg("+nightly")
+                    .arg("rustc")
+                    .arg("--manifest-path")
+                    .arg(manifest_path)
+                    .arg("--lib")
+                    .arg("--target")
+                    .arg("nvptx64-nvidia-cuda")
+                    .arg("--release")
+                    .arg("--")
+                    .arg("--emit")
+                    .arg("asm")
+                    .output()
+                    .map_err(|e| format!("Failed to execute command: {}", e))?;
+
+                if output.status.success() {
+                    let ptx_path = project_dir.join("target/nvptx64-nvidia-cuda/release");
+                    if ptx_path.exists() {
+                        let ptx_files: Vec<Ptx> = fs::read_dir(ptx_path)
+                        .map_err(|e| e.to_string())?
+                        .filter_map(|entry| {
+                            let path = entry.unwrap().path();
+                            if let Some("ptx") = path.extension().and_then(|ext| ext.to_str()) {
+                                Some(Ptx::from_file(path))
```

Comment:
Let's change this to
1. Read file contents using std::fs
2. use `Ptx::from_src` with the contents
Then we don't have to worry about files being removed invalidating the Ptx objects.